test/distributed/org/apache/cassandra/distributed/impl/InstanceAwareNodetool.java

```diff
@@ -0,0 +1,264 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.distributed.impl;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import org.junit.Assert;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.distributed.Cluster;
+import org.apache.cassandra.distributed.api.IInvokableInstance;
+import org.apache.cassandra.distributed.api.IIsolatedExecutor;
+import org.apache.cassandra.distributed.mock.nodetool.InternalNodeProbeFactory;
+import org.apache.cassandra.distributed.shared.AbstractBuilder;
+import org.apache.cassandra.tools.INodeProbeFactory;
+import org.apache.cassandra.tools.NodeTool;
+import org.apache.cassandra.tools.ToolRunner;
+
+/**
+ * Handy class for invoking nodetool against an instance, the difference between this
+ * and nodetool method invocation on IInstance is that we can have an output from this one which is not
+ * possible on IInstance nodetool invocation.
+ * <p>
+ * This is the gateway for testing the output of nodetool command and we can do various asserts on that, e.g.
+ * if the output is of some format and that is not broken (even accidentally).
+ * <p>
+ * The actual usage is like:
+ *
+ * <pre>
+ * {@code
+ * public class GossipinfoTest extends TestBaseImpl
+ * {
+ *     @Test
+ *     public void testGossipinfo() throws Exception
+ *     {
+ *         try (final InstanceAwareNodetool nodetool = new InstanceAwareNodetool(getCluster()))
+ *         {
+ *             String[] results = nodetool.execute("gossipinfo");
+ *         }
+ *     }
+ * }
+ * }</pre>
+ */
+public class InstanceAwareNodetool implements IIsolatedExecutor.SerializableCallable<String[]>, AutoCloseable
```

Comment:
mixed feelings about this class.  For one its practical and solves the problem at hand, but on the other hand it doesn't compose with the existing nodetool testing feature...

I would love to see org.apache.cassandra.distributed.api.NodeToolResult updated with stdout/stderr.  Looking closer at `NodetoolWrapper` since it will just run in-process and this does basically the same as the existing logic, the stdout/err logs will have logs from the daemon as well, so simply reading stdout/err should be enough.

I feel like this class could go away in favor of the following

```
class NodeToolResultsWithOutput {
private final NodeToolResult result;
private final ByteArrayOutputStream stdout;
private final ByteArrayOutputStream stderr;

public NodeToolResultsWithOutput(NodeToolResult result, ByteArrayOutputStream stdout, ByteArrayOutputStream stderr) {
  this.result = result;
  this.stdout = stdout;
  this.stderr = stderr;
}

  public NodeToolResult getResult() { return result; }
  public String getStdout() { return stdout.toString(); }
  public String getStderr() { return stderr.toString(); }
}

public static NodeToolResultsWithOutput nodetool(IInvokeableInstanc inst, boolean withNotifications, String... commandAndArgs) {
  return inst.callOnInstance(() -> {
               PrintStream originalSysOut = System.out;
                PrintStream originalSysErr = System.err;
                InputStream originalSysIn = System.in;
                originalSysOut.flush();
                originalSysErr.flush();
                ByteArrayOutputStream toolOut = new ByteArrayOutputStream();
                ByteArrayOutputStream toolErr = new ByteArrayOutputStream();
                
                System.setIn(stdin == null ? originalSysIn : stdin);
                try (PrintStream newOut = new PrintStream(toolOut); PrintStream newErr = new PrintStream(toolErr);)
                {
                    System.setOut(newOut);
                    System.setErr(newErr);
                    DTestNodeTool nodetool = new DTestNodeTool(withNotifications);
                    int rc =  nodetool.execute(commandAndArgs);
                    NodeToolResult results = NodeToolResult(commandAndArgs, rc, new ArrayList<>(nodetool.notifications.notifications), nodetool.latestError);
                    return new NodeToolResultsWithOutput(results, toolOut, toolErr);
                } finally {
                     System.setOut(originalSysOut);
                     System.setErr(originalSysErr);
                     System.setIn(originalSysIn);
                }
  });
}
```

TL;DR - long term I feel this logic should be pulled into instance.nodetoolWithResults, short term can we do something like the above so the logic can be used in other places?

Reply:
@dcapwell the amount of changes in order to rewrite it like you want is just not worth it. I would need to abstract this into some kind of wrapper which would call tool repeatedly, gather results etc, it is just additional work to rewrite it just for ... what exactly?

Reply:
spoke in slack, gave what I was asking for and tested it works.

The issue I see here is that

1) doesn't handle notifications, so any test which wants logs + notifications can't use
2) returns String[] which isn't clear how to use without reading the code; returning a structured object solves this
3) close closes cluster, but multiple nodetool commands are common.
4) half the methods duplicate NodetoolResult, so can remove if returning that.
5) timeout doesn't do anything since you are using a class for ToolRunner, so it blocks waiting for it to exit, so will never timeout

> it is just additional work to rewrite it just for ... what exactly?

nodetool stdout is a common need, so making this accessible to other use cases is a big win.  Also, what I was asking for is smaller than this (27 lines of code (plus 33 for the pojo) vs 264 lines.